src/coreclr/nativeaot/System.Private.CoreLib/src/System/InvokeUtils.cs

```diff
@@ -134,87 +134,289 @@ private static Exception ConvertOrWidenPrimitivesEnumsAndPointersIfPossible(obje
+                    break;
+
                 case EETypeElementType.UInt64:
-                    ulong ulongValue = Convert.ToUInt64(srcObject);
+                    ulong ulongValue;
+                    switch (srcElementType)
+                    {
+                        case EETypeElementType.UInt64:
+                            ulongValue = (ulong)srcObject;
+                            break;
+                        case EETypeElementType.Single:
+                            ulongValue = (ulong)(float)srcObject;
+                            break;
+                        case EETypeElementType.Double:
+                            ulongValue = (ulong)(double)srcObject;
+                            break;
+                        default:
+                            goto Failure;
+                    }
                     dstObject = dstEEType->IsEnum ? Enum.ToObject(dstEEType, (long)ulongValue) : ulongValue;
                     break;
 
                 case EETypeElementType.Single:
-                    if (new EETypePtr(srcEEType).CorElementType == CorElementType.ELEMENT_TYPE_CHAR)
+                    if (srcElementType == EETypeElementType.Single)
                     {
-                        dstObject = (float)(char)srcObject;
+                        dstObject = (float)srcObject;
                     }
                     else
                     {
-                        dstObject = Convert.ToSingle(srcObject);
+                        // Can only be double here
+                        dstObject = (float)(double)srcObject;
                     }
                     break;
 
                 case EETypeElementType.Double:
-                    if (new EETypePtr(srcEEType).CorElementType == CorElementType.ELEMENT_TYPE_CHAR)
-                    {
-                        dstObject = (double)(char)srcObject;
-                    }
-                    else
-                    {
-                        dstObject = Convert.ToDouble(srcObject);
-                    }
+
+                    // Can only be double here
+                    dstObject = (double)srcObject;
                     break;
 
                 default:
-                    Debug.Fail("Unexpected CorElementType: " + dstElementType + ": Not a valid widening target.");
-                    dstObject = null;
-                    return CreateChangeTypeException(srcEEType, dstEEType, semantics);
+                    goto Failure;
             }
 
             Debug.Assert(dstObject.GetMethodTable() == dstEEType);
             return null;
+
+            Failure:
+            Debug.Fail("Unexpected CorElementType: " + dstElementType + ": Not a valid widening target.");
+            dstObject = null;
+            return CreateChangeTypeException(srcEEType, dstEEType, semantics);
         }
 
         private static bool CanPrimitiveWiden(EETypeElementType destType, EETypeElementType srcType)
```

Comment:
Delete - no longer used?

Reply:
Removed in 54092d46e9cc15a75cb7afaaeaaf49b4d863b67d.

Reply:
Can we delete the `CanPrimitiveWiden` method and create the exception at the end? I do not think that we care about how fast the exception gets created.

Reply:
I deleted it in 95ea639b37843e88b7247eb77539556ebab88f39 🙂
Will leave this open, feel free to resolve if that change matches what you had in mind.